Implement TOTP for 2FA

mixonic opened the pull request:
This PR implements TOTP one time password support for 2FA in Brave.

A screencast walkthrough of this feature (6:11): https://youtu.be/znLjP6PfJzk

A technical summary:

* Add [ROTP](https://github.com/mdp/rotp) gem to support the logic for TOTP
* Add [rqrcode](https://github.com/whomwah/rqrcode) gem to support rendering a QR code as an SVG
* We're making some judgement calls about authentication (not all implemented, will check off when they are):
  * [x] When a user with an email link, TOTP, and U2F logs in they are prompted for U2F but they can opt-in to using TOTP for that auth attempt. They *must* use U2F or TOTP.
  * [x] When a user with an email link and U2F logs in the *must* use U2F.
  * [x] When a user with an email link and TOTP logs in they *must* use TOTP.
  * [x] When a user with an email link and no 2FA logs in the link itself is sufficient.
  * [x] When a user logs in via OAuth no 2FA is required even if it is configured (Perhaps those users should not be permitted to register 2FA).
* A user can have only one TOTP registration since a single registration (secret) is permitted to be hooked into several devices. For example, at registration time you could configure two phones to have secret then store one securely. If you lost your commonly used device, the backup device can be used to log in. Then if you wanted to revoke access for the lost device, you may generate a new registration (secret) to replace the existing one on your account.
* This implementation verifies TOTP passwords with +/- 30 seconds of drift permitted.
* This implementation [prevents re-use of TOTP passwords](https://github.com/mdp/rotp#preventing-reuse-of-time-based-otps) by tracking the last verified login time.

Resolves https://github.com/brave-intl/publishers/issues/363

Submitter Checklist:

- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Handle the case of a user with both TOTP and U2F registered logging in. Also when that user is on a browser that doesn't support U2F.
  - [x] better tests for this
  - [x] proper string translations
- [x] Unify the U2F and TOTP dashboards into a single page.
- [x] Deletion flow to revoke a TOTP registration.
- [x] Sign off by @jenn-rhim 
- [x] Sign off by @ayumi 

Reviewer Checklist:

Tests
- [ ] Adequate test coverage exists to prevent regressions

Security:
- [ ] No raw SQL -- Always prefer ActiveRecord query helpers ([more info on StackOverflow](https://stackoverflow.com/questions/41410752/rails-5-sql-injection#41452695))
- [ ] XSS is mitigated -- Avoid `html_safe` and `raw`; escape untrusted content from users and 3rd party APIs ([Rails XSS guide](https://brakemanpro.com/2017/09/08/cross-site-scripting-in-rails) and [OWASP XSS Prevention Cheat Sheet](https://www.owasp.org/index.php/XSS_(Cross_Site_Scripting)_Prevention_Cheat_Sheet))

mixonic commented:
@jenn-rhim @ayumi We think this is ready for a detailed security review, integration into the staging branch, and testing and experimentation by Brave staff on staging and production "dark launch". I'd like to spin off further design implementation in other PRs.

ayumi commented on app/views/two_factor_authentications/_totp.html.slim:
```diff
@@ -0,0 +1,19 @@
+div.text-center
+  = form_tag totp_authentications_path do
+    div.form-group
+      input name="totp_password" placeholder=t(".enter_code_placeholder") class="form-control"
```
Minor: consider setting `autofocus` here for improved UX

ayumi commented on app/views/totp_registrations/new.html.slim:
```diff
@@ -0,0 +1,34 @@
+                      .totp-svg== qr_code_svg @provisioning_url
+                  li.form-group
+                    = t '.step_3'
+                    input.form-control name="totp_password" placeholder=t('.password_prompt')
```
minor: consider autofocus here for improved UX :)

kpfefferle commented on test/fixtures/totp_registrations.yml:
```diff
@@ -0,0 +1,15 @@
+default:
+  publisher: verified
+  last_logged_in_at: <%= 2.days.ago %>
+  encrypted_secret_iv: "<%= Base64.encode64(TotpRegistration.encryption_key) %>"
```
This is following the same fixture pattern we also used for encrypted attributes in the Publisher fixture:

https://github.com/brave-intl/publishers/blob/fa7c5e4ccb1b009a7a171a5e320065c810939135/test/fixtures/publishers.yml#L31-L36

Do we want to update both fixtures to a different pattern?

cc: @mixonic

mixonic replied:
@ayumi's suggestion here is only that we should not be using the encryption_key for salt, which I think is reasonable. We can try to generate a random salt but would need to store it in a variable to be re-used in both the attribute field and in the encryption step. If that isn't possible we could fall back to `'fake_salt'` or something. I'll take a pass at it.

mixonic commented:
Great, @ayumi we've addressed your feedback here and have sign off for getting this into staging from @jenn-rhim via Slack. She opened https://github.com/brave-intl/publishers/issues/395 with further UI iteration that we can take on pretty much right after this.

Good to merge, I believe!
